### PR TITLE
Optimize Regex.escape/1

### PR DESCRIPTION
**Pull request by michalmuskala**
Use a similar optimization technique to the one in plug https://github.com/elixir-lang/plug/pull/547

This allows us to keep as large parts of the initial binary as possible,
in the case of no escapes, no allocation is needed.

In most cases it should be about twice as fast, in case no escape is needed,
in can be even 10x faster. For small strings, it will be about 15% slower, but
with entire execution in range of 50μs, this should be acceptable.

### CODE REVIEW

File: lib/elixir/lib/regex.ex [hunk 1]
```diff
@@ -714,14 +714,30 @@ defmodule Regex do
   """
   @spec escape(String.t) :: String.t
   def escape(string) when is_binary(string) do
-    escape(string, [])
+    IO.iodata_to_binary(escape(string, 0, string))
```
Comment (whatyouhide) [inline lib/elixir/lib/regex.ex]: I think something like

```elixir
string
|> escape(_length = 0, string)
|> IO.iodata_to_binary()
```

reads better here, thoughts?

File: lib/elixir/lib/regex.ex [hunk 2]
```diff
@@ -714,14 +714,30 @@ defmodule Regex do
...
   end
 
-  for char <- '.^$*+?()[]{}|#-\\\t\n\v\f\r\s' do
-    defp escape(<<unquote(char), rest::binary>>, acc), do: escape(rest, [acc, ?\\, unquote(char)])
+  @escapable '.^$*+?()[]{}|#-\\\t\n\v\f\r\s'
+
+  defp escape(<<char, rest::binary>>, len, original) when char in @escapable do
+    escape_char(rest, len, original, char)
```
Comment (whatyouhide) [inline lib/elixir/lib/regex.ex]: I am not a fan of the a-b recursion. Can we do

```elixir
case length do
  0 ->
    [?\\, char | escape(rest, 0, rest)]
  _other ->
    [binary_part(original, 0, length), ?\\, char | escape(rest, 0, rest)]
end
```

?

File: lib/elixir/lib/regex.ex [hunk 3]
```diff
@@ -714,14 +714,30 @@ defmodule Regex do
...
   def escape(string) when is_binary(string) do
-    escape(string, [])
+    IO.iodata_to_binary(escape(string, 0, string))
   end
 
-  for char <- '.^$*+?()[]{}|#-\\\t\n\v\f\r\s' do
-    defp escape(<<unquote(char), rest::binary>>, acc), do: escape(rest, [acc, ?\\, unquote(char)])
+  @escapable '.^$*+?()[]{}|#-\\\t\n\v\f\r\s'
+
+  defp escape(<<char, rest::binary>>, len, original) when char in @escapable do
```
Comment (whatyouhide) [inline lib/elixir/lib/regex.ex]: You're calling sometimes `len` and sometimes `length`. Let's go with `length` in every place.

File: lib/elixir/lib/regex.ex (hunk 2 above)
Comment (michalmuskala) [inline lib/elixir/lib/regex.ex]: No we can't. We'll loose binary optimisation in that case since we're using `rest` twice in the body.

Comment (whatyouhide) [inline lib/elixir/lib/regex.ex]: Right. Sucks :(

Comment (whatyouhide) [inline lib/elixir/lib/regex.ex]: Let's do the other changes then and it's good to go.

**Comment by michalmuskala** [comment]
@whatyouhide updated

**Pull request merged**